viadot/sources/cloud_for_customers.py

```diff
@@ -5,56 +5,169 @@
+            elif "my341115" in direct_url:
+                access = credentials.get("Prod")
+        elif url:
+            if "my336539" in url:
+                access = credentials.get("QA")
+            elif "my341115" in url:
+                access = credentials.get("Prod")
+        self.auth = (access["username"], access["password"])
+        self.direct_url = direct_url
 
     def to_records(self) -> List:
+        entity_list = []
+        first = True
+        while True:
+            if self.direct_url:
+                url = self.direct_url
+                response = self.check_url(url)
+            elif self.url:
+                if first:
+                    url = urljoin(self.url, self.query_endpoint)
+                    response = self.check_url(url, params=self.params)
+                    first = False
+                else:
+                    response = self.check_url(url)
+
+            if response.status_code == 200:
+                dirty_json = response.json()
+                entity_from_response = response_to_entity_list(dirty_json, url)
+                entity_list.append(entity_from_response)
+            url = dirty_json["d"].get("__next")
+            if url is None:
+                break
+        flat_entity_list = [item for sublist in entity_list for item in sublist]
+        return flat_entity_list
+
+    def check_url(self, url, params=None):
         try:
-            response = requests.get(
-                urljoin(self.api_url, self.query_endpoint),
-                params=self.params,
-                auth=self.auth,
+            session = requests.Session()
+            retry_strategy = Retry(
+                total=3,
+                status_forcelist=[429, 500, 502, 503, 504],
+                backoff_factor=1,
             )
+            adapter = HTTPAdapter(max_retries=retry_strategy)
+
+            session.mount("http://", adapter)
+            session.mount("https://", adapter)
+            response = session.get(url, params=params, auth=self.auth)
             response.raise_for_status()
+        except ReadTimeout as e:
+            msg = "The connection was successful, "
+            msg += f"however the API call to {url} timed out after .... s "
+            msg += "while waiting for the server to return data."
+            raise APIError(msg)
+
+        except HTTPError as e:
+            raise APIError(
+                f"The API call to {url} failed. "
+                "Perhaps your account credentials need to be refreshed?",
+            ) from e
 
+        except (ConnectionError, Timeout) as e:
+            raise APIError(
+                f"The API call to {url} failed due to connection issues."
+            ) from e
+        except ProtocolError as e:
+            raise APIError(f"Did not receive any reponse for the API call to {url}.")
         except Exception as e:
-            raise
+            raise APIError("Unknown error.") from e
```

Comment:
Since we're doing it for the second time, we should think about abstracting this whole procedure of handling the response into a function inside `utils.py`. I don't expect you to be able to do this yet, but I think you should leave a comment about this here (eg. # TODO: abstract below and put as `handle_api_response()` into `utils.py`). There should also be an issue raised for this.